apps/cuda_mat_mul/mat_mul_generator.cpp

```diff
@@ -40,14 +49,9 @@ class MatMul : public Halide::Generator<MatMul> {
             .compute_at(prod, y)
             .vectorize(B.in().args()[0]);
 
-        OutputImageParam bufs[] = {A, B, prod.output_buffer()};
```

Comment:
Is there no suitable base class for input and output? Constrainable?

Reply:
Well:
- Constrainable is an abstract class, so we can't use `Constrainable bufs[]`
- arrays-of-reference are illegal in C++, so we can't use `Constrainable &bufs[]`
- We could of course use pointers to do something like
```
        Halide::Internal::Constrainable *bufs[] = {&A, &B, &out};
        for (auto *buf : bufs) {
            buf->set_host_alignment(16)
                .dim(0).set_bounds(0, size)       
                .dim(1).set_stride(size);     
         }     
```
But it's a little non-idiomatic in Halide. (And of course Constrainable is currently Internal but that could be fixed.)

That said, Constrainable could probably be made to be a non-abstract value class with a little bit of reworking, but it's probably better served in a separate PR.

How about I land this as-is, then revisit in a subsequent PR to make Constrainable more useful?

Reply:
sgtm